File: python/Advanced Graphs/alien_dictionary.py
from typing import List

class Solution:
    def foreignDictionary(self, words: List[str]) -> str:
        """
        Given a list of words sorted lexicographically in an alien language, return a string of the unique characters in the correct order.
        If the order is invalid (cycle or prefix issue), return an empty string.

        Approach:
        1. Build a graph (adjacency list) where each node is a character, and edges represent precedence (w1[j] < w2[j]).
        2. For each adjacent pair of words, find the first differing character and add an edge.
        3. Detect invalid prefix case (e.g., ["abc", "ab"]).
        4. Perform DFS-based topological sort to determine character order.
        5. Detect cycles (invalid order) during DFS.

        Time Complexity: O(N + K)
            - N = total number of characters in all words
            - K = number of unique characters (nodes in graph)
            - Building the graph: O(N)
            - DFS/topological sort: O(K + E), E = number of edges (<= K^2)
        Space Complexity: O(K + E)
            - For adjacency list, visit dict, and recursion stack
        """
        # Special case: single word, return unique characters in order of appearance
        if len(words) == 1:
            return "".join(dict.fromkeys(words[0]))

        adj = {c : set() for w in words for c in w}  # Build graph nodes for all unique characters

        # Build edges based on the first differing character between adjacent words
        for i in range(len(words) - 1):
            w1, w2 = words[i], words[i + 1]
            minLen = min(len(w1), len(w2))
            # Invalid case: prefix order (e.g., "abc" before "ab")
            if len(w1) > len(w2) and w1[: minLen] == w2[: minLen]:
                return ""
            for j in range(minLen):
                if w1[j] != w2[j]:
                    adj[w1[j]].add(w2[j])  # w1[j] comes before w2[j]
                    break

        visit = {} # False = visited, True = current path (for cycle detection)
        res = []   # Stores the topological order
        
        def dfs(c):
            if c in visit:
                return visit[c]  # True if cycle detected
            
            visit[c] = True  # Mark as visiting (current path)

            for nei in adj[c]:
                if dfs(nei):  # If cycle detected in neighbor
                    return True

            visit[c] = False  # Mark as visited (no cycle in this path)
            res.append(c)     # Add to result in postorder (topological sort)

        # Run DFS for all nodes
        for c in adj:
            if dfs(c):  # If cycle detected
                return ""
        res.reverse()  # Reverse to get correct order
        return "".join(res)


def test_foreign_dictionary():
    solution = Solution()
    # Test case 1: Simple order
    words1 = ["wrt", "wrf", "er", "ett", "rftt"]
    expected1 = "wertf"
    result1 = solution.foreignDictionary(words1)
    print(f"Test 1: {result1 == expected1}, Expected: {expected1}, Got: {result1}")

    # Test case 2: Cycle (invalid order)
    words2 = ["z", "x", "z"]
    expected2 = ""
    result2 = solution.foreignDictionary(words2)
    print(f"Test 2: {result2 == expected2}, Expected: {expected2}, Got: {result2}")

    # Test case 3: Prefix invalid
    words3 = ["abc", "ab"]
    expected3 = ""
    result3 = solution.foreignDictionary(words3)
    print(f"Test 3: {result3 == expected3}, Expected: {expected3}, Got: {result3}")

    # Test case 4: Single word
    words4 = ["abc"]
    expected4 = "abc"
    result4 = solution.foreignDictionary(words4)
    print(f"Test 4: {result4 == expected4}, Expected: {expected4}, Got: {result4}")

    # Test case 5: All characters same
    words5 = ["aaa", "aaa"]
    expected5 = "a"
    result5 = solution.foreignDictionary(words5)
    print(f"Test 5: {result5 == expected5}, Expected: {expected5}, Got: {result5}")

if __name__ == "__main__":
    test_foreign_dictionary()

File: C++/Advanced Graphs/alien_dictionary.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    string foreignDictionary(vector<string>& words) {
        /*
        Approach:
        - Build a graph (adjacency list) where each character points to characters that must come after it.
        - Compare adjacent words: find the first differing character and add an edge.
        - Detect invalid case: when a longer word comes before its prefix.
        - Perform DFS-based topological sort to get character order.
        - Detect cycles (invalid ordering) during DFS.
        
        Time Complexity: O(N + K + E)
            - N = total number of characters in all words
            - K = number of unique characters
            - E = number of edges (<= K^2)
            - Build graph: O(N)
            - Topological sort: O(K + E)

        Space Complexity: O(K + E)
            - For graph, visited map, result, and recursion stack.
        */

        unordered_map<char, unordered_set<char>> adj;  // adjacency list
        unordered_map<char, int> visit;  // 0 = unvisited, 1 = visiting, 2 = visited
        string res;

        // Create nodes for all unique characters
        for (string& w : words) {
            for (char c : w) {
                adj[c]; // ensures node exists
            }
        }

        // Build edges from adjacent words
        for (int i = 0; i < words.size() - 1; ++i) {
            string w1 = words[i], w2 = words[i + 1];
            int minLen = min(w1.size(), w2.size());
            if (w1.size() > w2.size() && w1.substr(0, minLen) == w2.substr(0, minLen)) {
                return ""; // invalid prefix case
            }

            for (int j = 0; j < minLen; ++j) {
                if (w1[j] != w2[j]) {
                    adj[w1[j]].insert(w2[j]); // w1[j] < w2[j]
                    break;
                }
            }
        }

        // DFS function
        function<bool(char)> dfs = [&](char c) {
            if (visit[c] == 1) return true;  // cycle detected
            if (visit[c] == 2) return false; // already visited
            
            visit[c] = 1; // mark as visiting

            for (char nei : adj[c]) {
                if (dfs(nei)) return true;
            }

            visit[c] = 2; // mark as visited
            res.push_back(c);
            return false;
        };

        // Run DFS for all nodes
        for (auto& p : adj) {
            char c = p.first;
            if (!visit[c]) {
                if (dfs(c)) return ""; // cycle detected
            }
        }

        reverse(res.begin(), res.end()); // reverse to get correct order
        return res;
    }
};

/*
================== LINE BY LINE EXPLANATION ==================

#include <bits/stdc++.h>
- Includes all standard C++ libraries.

using namespace std;
- Avoids writing std:: repeatedly.

class Solution {
- Defines the Solution class.

string foreignDictionary(vector<string>& words) {
- Function to find the alien dictionary order.

unordered_map<char, unordered_set<char>> adj;
- Graph: maps each char to a set of chars that must come after it.

unordered_map<char, int> visit;
- Track state of each node: 0 = unvisited, 1 = visiting, 2 = visited.

string res;
- To store the final order of characters.

for (string& w : words) { for (char c : w) adj[c]; }
- Ensures all unique characters appear as nodes in graph.

for (int i = 0; i < words.size()-1; ++i) {
- Compare adjacent word pairs.

if (w1.size() > w2.size() && w1.substr(0, minLen) == w2.substr(0, minLen))
- If longer word comes before its prefix: invalid case.

if (w1[j] != w2[j]) { adj[w1[j]].insert(w2[j]); break; }
- Add edge w1[j] → w2[j] for first differing char.

function<bool(char)> dfs = [&](char c) {
- Lambda for DFS traversal to detect cycles and perform topo sort.

if (visit[c] == 1) return true;
- If currently visiting this node: cycle detected.

if (visit[c] == 2) return false;
- If already fully processed: skip.

visit[c] = 1;
- Mark as visiting.

for (char nei : adj[c]) { if (dfs(nei)) return true; }
- Visit all neighbors recursively.

visit[c] = 2;
- Mark as fully processed.

res.push_back(c);
- Add to result after processing all neighbors.

for (auto& [c, _] : adj) {
- Run DFS for all unvisited nodes.

if (dfs(c)) return "";
- If cycle detected, return empty string.

reverse(res.begin(), res.end());
- Reverse the result to get the correct order.

return res;
- Return final string.

================================================================
*/

